### PR TITLE
Bench `base58` encoding and remove `SmallVec` to improve perf

### PR DESCRIPTION
**Pull request by RCasatta**
In a downstream app I've seen printing a descriptor is not a cheap operation, analyzing the flamegraph it seems base58 encoding of the xpub is the culprit

image

This PR adds benches for the `encode_check` function, and add the changes gaining more boost, which is also good cause it removes code.

Other attempts didn't provide enough benefit for inclusion but I report them here for knowledge.

```
## baseline

running 2 tests
test benches::bench_encode_check_50   ... bench:       8,760 ns/iter (+/- 113)
test benches::bench_encode_check_xpub ... bench:      19,982 ns/iter (+/- 109)


## remove smallvec

running 2 tests
test benches::bench_encode_check_50   ... bench:       7,935 ns/iter (+/- 129)
test benches::bench_encode_check_xpub ... bench:      18,076 ns/iter (+/- 184)


## increase smallvec to 128 (fits xpub)

test benches::bench_encode_check_50   ... bench:       8,786 ns/iter (+/- 738)
test benches::bench_encode_check_xpub ... bench:      20,006 ns/iter (+/- 2,611)


## avoid char-to-str by keeping str map

test benches::bench_encode_check_50   ... bench:       7,895 ns/iter (+/- 88)
test benches::bench_encode_check_xpub ... bench:      17,883 ns/iter (+/- 118)
```

Gains are good (~10%), but I don't think they explains the 3ms to print a descriptor in wasm env, 
probably is the sha256 for the checksum is fast in cargo bench but slow in wasm env, but I didn't research on the topic.

### CODE REVIEW

**Comment by apoelstra** [comment]
utACK deeb160b86e5e025ea98494b5d30611e47dc8769 -- I also see a 10% improvement and the `base58` crate is already inundated with `Vec`s and requires an allocator. (Maybe we were hoping to move to a smallvec-based scheme eventually and drop the allocator, but there's lots of work to do there and base58 is not where I'd start..)

However, amusingly you have unearthed a clippy warning:
```
error: it looks like the same item is being pushed into this Vec
   --> base58/src/lib.rs:182:9
    |
182 |         ret.push(0);
    |         ^^^
    |
    = help: consider using vec![0;SIZE] or ret.resize(NEW_SIZE, 0)
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#same_item_push
    = note: `-D clippy::same-item-push` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::same_item_push)]`
 
error: could not compile `base58ck` (lib) due to 1 previous error
```

This is not in the code you directly touched, but it did not trigger before because it is `Vec`-specific and we used to use `SmallVec`.

**Review by apoelstra — APPROVED** [review]
ACK 46466905211a35d82033e100d0240d332145aaef

File: base58/src/lib.rs [hunk 1]
```diff
@@ -115,7 +118,9 @@ pub fn decode_check(data: &str) -> Result<Vec<u8>, Error> {
 }
 
 /// Encodes `data` as a base58 string (see also `base58::encode_check()`).
-pub fn encode(data: &[u8]) -> String { encode_iter(data.iter().cloned()) }
+pub fn encode(data: &[u8]) -> String {
+    encode_iter(data.iter().cloned())
+}
```
Comment (tcharding) [inline base58/src/lib.rs:123]: nit: This is an unrelated whitespace change, and introduces a formating issue.

**Review by tcharding — APPROVED** [review]
ACK 46466905211a35d82033e100d0240d332145aaef

File: base58/src/lib.rs [hunk 2]
```diff
@@ -187,37 +189,6 @@ where
     Ok(())
 }
 
-/// Vector-like object that holds the first 100 elements on the stack. If more space is needed it
-/// will be allocated on the heap.
-struct SmallVec<T> {
-    len: usize,
-    stack: [T; 100],
```
[the remark this answers is not in the document]
Comment (apoelstra) [reply base58/src/lib.rs:191]: Thanks for looking into this!

Yeah, I was pretty surprised to see a performance change in this direction but didn't take the time to dig into it.
